File: apps/sdkwork-appstore-flutter-mobile/packages/sdkwork_appstore_flutter_mobile_publisher/lib/src/state/publisher_state.dart
import '../models/publisher_models.dart';

enum PublisherStatus { idle, loading, ready, error }

class PublisherState {
  const PublisherState({
    this.status = PublisherStatus.idle,
    this.lastResult,
    this.error,
  });

  final PublisherStatus status;
  final PublisherPageResult<Object?>? lastResult;
  final String? error;

  PublisherState loading() => const PublisherState(status: PublisherStatus.loading);

  PublisherState loaded(PublisherPageResult<Object?> result) =>
      PublisherState(status: PublisherStatus.ready, lastResult: result);

  PublisherState failed(String message) =>
      PublisherState(status: PublisherStatus.error, error: message);
}

File: apps/sdkwork-appstore-mini-program/packages/sdkwork-appstore-mp-publisher/src/state/publisherState.ts
import type { PublisherPageResult } from "../types/publisherModels.js";

export interface PublisherState {
  readonly status: "idle" | "loading" | "ready" | "error";
  readonly lastResult?: PublisherPageResult<unknown>;
  readonly error?: string;
}

export function initialPublisherState(): PublisherState {
  return { status: "idle" };
}

export function reducePublisherState(
  state: PublisherState,
  event:
    | { type: "load" }
    | { type: "loaded"; result: PublisherPageResult<unknown> }
    | { type: "failed"; error: string },
): PublisherState {
  switch (event.type) {
    case "load":
      return { ...state, status: "loading", error: undefined };
    case "loaded":
      return { status: "ready", lastResult: event.result };
    case "failed":
      return { ...state, status: "error", error: event.error };
  }
}
